File: libs/moqtail-rs/src/model/control/constant.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use core::convert::From;

use crate::model::error::ParseError;

pub const DRAFT_11: u32 = 0xFF00000B;
pub const DRAFT_14: u32 = 0xFF00000E;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u64)]
pub enum ControlMessageType {
  ClientSetup = 0x20,
  ServerSetup = 0x21,
  GoAway = 0x10,
  MaxRequestId = 0x15,
  RequestsBlocked = 0x1A,
  Subscribe = 0x03,
  SubscribeOk = 0x04,
  SubscribeError = 0x05,
  Unsubscribe = 0x0A,
  SubscribeUpdate = 0x02,
  Fetch = 0x16,
  FetchOk = 0x18,
  FetchError = 0x19,
  FetchCancel = 0x17,
  TrackStatus = 0x0D,
  TrackStatusOk = 0x0E,
  TrackStatusError = 0x0F,
  PublishNamespace = 0x06,
  PublishNamespaceOk = 0x07,
  PublishNamespaceError = 0x08,
  PublishNamespaceDone = 0x09,
  PublishNamespaceCancel = 0x0C,
  SubscribeNamespace = 0x11,
  SubscribeNamespaceOk = 0x12,
  SubscribeNamespaceError = 0x13,
  UnsubscribeNamespace = 0x14,
  Publish = 0x1D,
  PublishDone = 0x0B,
  PublishOk = 0x1E,
  PublishError = 0x1F,
}

impl TryFrom<u64> for ControlMessageType {
  type Error = ParseError;

  fn try_from(value: u64) -> Result<Self, Self::Error> {
    match value {
      0x20 => Ok(ControlMessageType::ClientSetup),
      0x21 => Ok(ControlMessageType::ServerSetup),
      0x10 => Ok(ControlMessageType::GoAway),
      0x15 => Ok(ControlMessageType::MaxRequestId),
      0x1A => Ok(ControlMessageType::RequestsBlocked),
      0x03 => Ok(ControlMessageType::Subscribe),
      0x04 => Ok(ControlMessageType::SubscribeOk),
      0x05 => Ok(ControlMessageType::SubscribeError),
      0x0A => Ok(ControlMessageType::Unsubscribe),
      0x02 => Ok(ControlMessageType::SubscribeUpdate),
      0x0B => Ok(ControlMessageType::PublishDone),
      0x16 => Ok(ControlMessageType::Fetch),
      0x18 => Ok(ControlMessageType::FetchOk),
      0x19 => Ok(ControlMessageType::FetchError),
      0x17 => Ok(ControlMessageType::FetchCancel),
      0x0D => Ok(ControlMessageType::TrackStatus),
      0x0E => Ok(ControlMessageType::TrackStatus),
      0x06 => Ok(ControlMessageType::PublishNamespace),
      0x07 => Ok(ControlMessageType::PublishNamespaceOk),
      0x08 => Ok(ControlMessageType::PublishNamespaceError),
      0x09 => Ok(ControlMessageType::PublishNamespaceDone),
      0x0C => Ok(ControlMessageType::PublishNamespaceCancel),
      0x11 => Ok(ControlMessageType::SubscribeNamespace),
      0x12 => Ok(ControlMessageType::SubscribeNamespaceOk),
      0x13 => Ok(ControlMessageType::SubscribeNamespaceError),
      0x14 => Ok(ControlMessageType::UnsubscribeNamespace),
      0x1D => Ok(ControlMessageType::Publish),
      0x1E => Ok(ControlMessageType::PublishOk),
      0x1F => Ok(ControlMessageType::PublishError),
      _ => Err(ParseError::InvalidType {
        context: " ControlMessageType::try_from(u64)",
        details: format!("Invalid type, got {value}"),
      }),
    }
  }
}

impl From<ControlMessageType> for u64 {
  fn from(value: ControlMessageType) -> Self {
    value as u64
  }
}

#[derive(Debug, Clone, Copy, PartialEq)]
#[repr(u64)]
pub enum FilterType {
  NextGroupStart = 0x1,
  LatestObject = 0x2,
  AbsoluteStart = 0x3,
  AbsoluteRange = 0x4,
}

impl TryFrom<u64> for FilterType {
  type Error = ParseError;

  fn try_from(value: u64) -> Result<Self, Self::Error> {
    match value {
      0x1 => Ok(FilterType::NextGroupStart),
      0x2 => Ok(FilterType::LatestObject),
      0x3 => Ok(FilterType::AbsoluteStart),
      0x4 => Ok(FilterType::AbsoluteRange),
      _ => Err(ParseError::InvalidType {
        context: "FilterType::try_from(u64)",
        details: format!("Invalid type, got {value}"),
      }),
    }
  }
}

impl From<FilterType> for u64 {
  fn from(value: FilterType) -> Self {
    value as u64
  }
}

#[derive(Debug, Clone, Copy, PartialEq)]
#[repr(u64)]
pub enum FetchType {
  StandAlone = 0x1,
  RelativeFetch = 0x2,
  AbsoluteFetch = 0x3,
}

impl TryFrom<u64> for FetchType {
  type Error = ParseError;

  fn try_from(value: u64) -> Result<Self, Self::Error> {
    match value {
      0x1 => Ok(FetchType::StandAlone),
      0x2 => Ok(FetchType::RelativeFetch),
      0x3 => Ok(FetchType::AbsoluteFetch),
      _ => Err(ParseError::InvalidType {
        context: "FetchType::try_from(u64)",
        details: format!("Invalid type, got {value}"),
      }),
    }
  }
}

impl From<FetchType> for u64 {
  fn from(value: FetchType) -> Self {
    value as u64
  }
}

#[derive(Debug, Clone, Copy, PartialEq)]
#[repr(u8)]
pub enum GroupOrder {
  Original = 0x0,
  Ascending = 0x1,
  Descending = 0x2,
}

impl TryFrom<u8> for GroupOrder {
  type Error = ParseError;

  fn try_from(value: u8) -> Result<Self, Self::Error> {
    match value {
      0x0 => Ok(GroupOrder::Original),
      0x1 => Ok(GroupOrder::Ascending),
      0x2 => Ok(GroupOrder::Descending),
      _ => Err(ParseError::InvalidType {
        context: "GroupOrder::try_from(u8)",
        details: format!("Invalid type, got {value}"),
      }),
    }
  }
}

impl From<GroupOrder> for u8 {
  fn from(value: GroupOrder) -> Self {
    value as u8
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u64)]
pub enum SubscribeErrorCode {
  InternalError = 0x0,
  Unauthorized = 0x1,
  Timeout = 0x2,
  NotSupported = 0x3,
  TrackDoesNotExist = 0x4,
  InvalidRange = 0x5,
  MalformedAuthToken = 0x10,
  ExpiredAuthToken = 0x12,
}

impl TryFrom<u64> for SubscribeErrorCode {
  type Error = ParseError;

  fn try_from(value: u64) -> Result<Self, Self::Error> {
    match value {
      0x0 => Ok(SubscribeErrorCode::InternalError),
      0x1 => Ok(SubscribeErrorCode::Unauthorized),
      0x2 => Ok(SubscribeErrorCode::Timeout),
      0x3 => Ok(SubscribeErrorCode::NotSupported),
      0x4 => Ok(SubscribeErrorCode::TrackDoesNotExist),
      0x5 => Ok(SubscribeErrorCode::InvalidRange),
      0x10 => Ok(SubscribeErrorCode::MalformedAuthToken),
      0x12 => Ok(SubscribeErrorCode::ExpiredAuthToken),
      _ => Err(ParseError::InvalidType {
        context: "SubscribeErrorCode::try_from(u64)",
        details: format!("Invalid type, got {value}"),
      }),
    }
  }
}

impl From<SubscribeErrorCode> for u64 {
  fn from(value: SubscribeErrorCode) -> Self {
    value as u64
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u64)]
pub enum FetchErrorCode {
  InternalError = 0x0,
  Unauthorized = 0x1,
  Timeout = 0x2,
  NotSupported = 0x3,
  TrackDoesNotExist = 0x4,
  InvalidRange = 0x5,
  NoObjects = 0x6,
  InvalidJoiningRequestId = 0x7,
  UnknownStatusInRange = 0x8,
  MalformedTrack = 0x9,
  MalformedAuthToken = 0x10,
  ExpiredAuthToken = 0x12,
}

impl TryFrom<u64> for FetchErrorCode {
  type Error = ParseError;

  fn try_from(value: u64) -> Result<Self, Self::Error> {
    match value {
      0x0 => Ok(FetchErrorCode::InternalError),
      0x1 => Ok(FetchErrorCode::Unauthorized),
      0x2 => Ok(FetchErrorCode::Timeout),
      0x3 => Ok(FetchErrorCode::NotSupported),
      0x4 => Ok(FetchErrorCode::TrackDoesNotExist),
      0x5 => Ok(FetchErrorCode::InvalidRange),
      0x6 => Ok(FetchErrorCode::NoObjects),
      0x7 => Ok(FetchErrorCode::InvalidJoiningRequestId),
      0x8 => Ok(FetchErrorCode::UnknownStatusInRange),
      0x9 => Ok(FetchErrorCode::MalformedTrack),
      0x10 => Ok(FetchErrorCode::MalformedAuthToken),
      0x12 => Ok(FetchErrorCode::ExpiredAuthToken),
      _ => Err(ParseError::InvalidType {
        context: "FetchErrorCode::try_from(u64)",
        details: format!("Invalid type, got {value}"),
      }),
    }
  }
}

impl From<FetchErrorCode> for u64 {
  fn from(value: FetchErrorCode) -> Self {
    value as u64
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u64)]
pub enum TrackStatusCode {
  InProgress = 0x00,
  DoesNotExist = 0x01,
  NotYetBegun = 0x02,
  Finished = 0x03,
  RelayUnavailable = 0x04,
}

impl TryFrom<u64> for TrackStatusCode {
  type Error = ParseError;

  fn try_from(value: u64) -> Result<Self, Self::Error> {
    match value {
      0x00 => Ok(TrackStatusCode::InProgress),
      0x01 => Ok(TrackStatusCode::DoesNotExist),
      0x02 => Ok(TrackStatusCode::NotYetBegun),
      0x03 => Ok(TrackStatusCode::Finished),
      0x04 => Ok(TrackStatusCode::RelayUnavailable),
      _ => Err(ParseError::InvalidType {
        context: "TrackStatusCode::try_from(u64)",
        details: format!("Invalid type, got {value}"),
      }),
    }
  }
}

impl From<TrackStatusCode> for u64 {
  fn from(value: TrackStatusCode) -> Self {
    value as u64
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u64)]
pub enum SubscribeNamespaceErrorCode {
  InternalError = 0x0,
  Unauthorized = 0x1,
  Timeout = 0x2,
  NotSupported = 0x3,
  NamespacePrefixUnknown = 0x4,
  NamespacePrefixOverlap = 0x5,
  MalformedAuthToken = 0x10,
  UnknownAuthTokenAlias = 0x11,
  ExpiredAuthToken = 0x12,
}
impl TryFrom<u64> for SubscribeNamespaceErrorCode {
  type Error = ParseError;

  fn try_from(value: u64) -> Result<Self, Self::Error> {
    match value {
      0x0 => Ok(SubscribeNamespaceErrorCode::InternalError),
      0x1 => Ok(SubscribeNamespaceErrorCode::Unauthorized),
      0x2 => Ok(SubscribeNamespaceErrorCode::Timeout),
      0x3 => Ok(SubscribeNamespaceErrorCode::NotSupported),
      0x4 => Ok(SubscribeNamespaceErrorCode::NamespacePrefixUnknown),
      0x5 => Ok(SubscribeNamespaceErrorCode::NamespacePrefixOverlap),
      0x10 => Ok(SubscribeNamespaceErrorCode::MalformedAuthToken),
      0x11 => Ok(SubscribeNamespaceErrorCode::UnknownAuthTokenAlias),
      0x12 => Ok(SubscribeNamespaceErrorCode::ExpiredAuthToken),
      _ => Err(ParseError::InvalidType {
        context: "SubscribeNamespaceErrorCode::try_from(u64)",
        details: format!("Invalid type, got {value}"),
      }),
    }
  }
}

impl From<SubscribeNamespaceErrorCode> for u64 {
  fn from(value: SubscribeNamespaceErrorCode) -> Self {
    value as u64
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u64)]
pub enum SubscribeDoneStatusCode {
  InternalError = 0x0,
  Unauthorized = 0x1,
  TrackEnded = 0x2,
  SubscriptionEnded = 0x3,
  GoingAway = 0x4,
  Expired = 0x5,
  TooFarBehind = 0x6,
}
impl TryFrom<u64> for SubscribeDoneStatusCode {
  type Error = ParseError;

  fn try_from(value: u64) -> Result<Self, Self::Error> {
    match value {
      0x0 => Ok(SubscribeDoneStatusCode::InternalError),
      0x1 => Ok(SubscribeDoneStatusCode::Unauthorized),
      0x2 => Ok(SubscribeDoneStatusCode::TrackEnded),
      0x3 => Ok(SubscribeDoneStatusCode::SubscriptionEnded),
      0x4 => Ok(SubscribeDoneStatusCode::GoingAway),
      0x5 => Ok(SubscribeDoneStatusCode::Expired),
      0x6 => Ok(SubscribeDoneStatusCode::TooFarBehind),
      _ => Err(ParseError::InvalidType {
        context: "SubscribeDoneStatusCode::try_from(u64)",
        details: format!("Invalid type, got {value}"),
      }),
    }
  }
}

impl From<SubscribeDoneStatusCode> for u64 {
  fn from(value: SubscribeDoneStatusCode) -> Self {
    value as u64
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u64)]
pub enum PublishNamespaceErrorCode {
  InternalError = 0x0,
  Unauthorized = 0x1,
  Timeout = 0x2,
  NotSupported = 0x3,
  Uninterested = 0x4,
  MalformedAuthToken = 0x10,
  UnknownAuthTokenAlias = 0x11,
  ExpiredAuthToken = 0x12,
}

impl TryFrom<u64> for PublishNamespaceErrorCode {
  type Error = ParseError;

  fn try_from(value: u64) -> Result<Self, Self::Error> {
    match value {
      0x0 => Ok(PublishNamespaceErrorCode::InternalError),
      0x1 => Ok(PublishNamespaceErrorCode::Unauthorized),
      0x2 => Ok(PublishNamespaceErrorCode::Timeout),
      0x3 => Ok(PublishNamespaceErrorCode::NotSupported),
      0x4 => Ok(PublishNamespaceErrorCode::Uninterested),
      0x10 => Ok(PublishNamespaceErrorCode::MalformedAuthToken),
      0x11 => Ok(PublishNamespaceErrorCode::UnknownAuthTokenAlias),
      0x12 => Ok(PublishNamespaceErrorCode::ExpiredAuthToken),
      _ => Err(ParseError::InvalidType {
        context: "PublishNamespaceErrorCode::try_from(u64)",
        details: format!("Invalid type, got {value}"),
      }),
    }
  }
}

impl From<PublishNamespaceErrorCode> for u64 {
  fn from(value: PublishNamespaceErrorCode) -> Self {
    value as u64
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u64)]
pub enum PublishDoneStatusCode {
  InternalError = 0x0,
  Unauthorized = 0x1,
  TrackEnded = 0x2,
  SubscriptionEnded = 0x3,
  GoingAway = 0x4,
  Expired = 0x5,
  TooFarBehind = 0x6,
  MalformedTrack = 0x7,
}

impl TryFrom<u64> for PublishDoneStatusCode {
  type Error = ParseError;

  fn try_from(value: u64) -> Result<Self, Self::Error> {
    match value {
      0x0 => Ok(PublishDoneStatusCode::InternalError),
      0x1 => Ok(PublishDoneStatusCode::Unauthorized),
      0x2 => Ok(PublishDoneStatusCode::TrackEnded),
      0x3 => Ok(PublishDoneStatusCode::SubscriptionEnded),
      0x4 => Ok(PublishDoneStatusCode::GoingAway),
      0x5 => Ok(PublishDoneStatusCode::Expired),
      0x6 => Ok(PublishDoneStatusCode::TooFarBehind),
      0x7 => Ok(PublishDoneStatusCode::MalformedTrack),
      _ => Err(ParseError::InvalidType {
        context: "PublishDoneStatusCode::try_from(u64)",
        details: format!("Invalid type, got {value}"),
      }),
    }
  }
}

impl From<PublishDoneStatusCode> for u64 {
  fn from(value: PublishDoneStatusCode) -> Self {
    value as u64
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u64)]
pub enum PublishErrorCode {
  InternalError = 0x0,
  Unauthorized = 0x1,
  Timeout = 0x2,
  NotSupported = 0x3,
  Uninterested = 0x4,
}

impl TryFrom<u64> for PublishErrorCode {
  type Error = ParseError;

  fn try_from(value: u64) -> Result<Self, Self::Error> {
    match value {
      0x0 => Ok(PublishErrorCode::InternalError),
      0x1 => Ok(PublishErrorCode::Unauthorized),
      0x2 => Ok(PublishErrorCode::Timeout),
      0x3 => Ok(PublishErrorCode::NotSupported),
      0x4 => Ok(PublishErrorCode::Uninterested),
      _ => Err(ParseError::InvalidType {
        context: "PublishErrorCode::try_from(u64)",
        details: format!("Invalid type, got {value}"),
      }),
    }
  }
}

impl From<PublishErrorCode> for u64 {
  fn from(value: PublishErrorCode) -> Self {
    value as u64
  }
}

File: libs/moqtail-ts/src/model/control/constant.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { CastingError } from '../error'
import { ClientSetup, ServerSetup } from '../control'
/**
 * 32 bit MOQT Draft-11 version number exchanged in {@link ClientSetup} and {@link ServerSetup}
 */
export const DRAFT_11 = 0xff00000b

/**
 * @public
 * Control message types for MOQT protocol.
 * Each value corresponds to a specific control frame.
 */
export enum ControlMessageType {
  ReservedSetupV00 = 0x01,
  ReservedClientSetupV10 = 0x40,
  ReservedServerSetupV10 = 0x41,
  ClientSetup = 0x20,
  ServerSetup = 0x21,
  GoAway = 0x10,
  MaxRequestId = 0x15,
  RequestsBlocked = 0x1a,
  Subscribe = 0x03,
  SubscribeOk = 0x04,
  SubscribeError = 0x05,
  Unsubscribe = 0x0a,
  SubscribeUpdate = 0x02,
  SubscribeDone = 0x0b,
  Fetch = 0x16,
  FetchOk = 0x18,
  FetchError = 0x19,
  FetchCancel = 0x17,
  TrackStatusRequest = 0x0d,
  TrackStatus = 0x0e,
  Announce = 0x06,
  AnnounceOk = 0x07,
  AnnounceError = 0x08,
  Unannounce = 0x09,
  AnnounceCancel = 0x0c,
  SubscribeAnnounces = 0x11,
  SubscribeAnnouncesOk = 0x12,
  SubscribeAnnouncesError = 0x13,
  UnsubscribeAnnounces = 0x14,
}

/**
 * Converts a bigint value to a ControlMessageType enum.
 * @param v - The bigint value.
 * @returns The corresponding ControlMessageType.
 * @throws Error if the value is not a valid control message type.
 */
export function controlMessageTypeFromBigInt(v: bigint): ControlMessageType {
  switch (v) {
    case 0x01n:
      return ControlMessageType.ReservedSetupV00
    case 0x40n:
      return ControlMessageType.ReservedClientSetupV10
    case 0x41n:
      return ControlMessageType.ReservedServerSetupV10
    case 0x20n:
      return ControlMessageType.ClientSetup
    case 0x21n:
      return ControlMessageType.ServerSetup
    case 0x10n:
      return ControlMessageType.GoAway
    case 0x15n:
      return ControlMessageType.MaxRequestId
    case 0x1an:
      return ControlMessageType.RequestsBlocked
    case 0x03n:
      return ControlMessageType.Subscribe
    case 0x04n:
      return ControlMessageType.SubscribeOk
    case 0x05n:
      return ControlMessageType.SubscribeError
    case 0x0an:
      return ControlMessageType.Unsubscribe
    case 0x02n:
      return ControlMessageType.SubscribeUpdate
    case 0x0bn:
      return ControlMessageType.SubscribeDone
    case 0x16n:
      return ControlMessageType.Fetch
    case 0x18n:
      return ControlMessageType.FetchOk
    case 0x19n:
      return ControlMessageType.FetchError
    case 0x17n:
      return ControlMessageType.FetchCancel
    case 0x0dn:
      return ControlMessageType.TrackStatusRequest
    case 0x0en:
      return ControlMessageType.TrackStatus
    case 0x06n:
      return ControlMessageType.Announce
    case 0x07n:
      return ControlMessageType.AnnounceOk
    case 0x08n:
      return ControlMessageType.AnnounceError
    case 0x09n:
      return ControlMessageType.Unannounce
    case 0x0cn:
      return ControlMessageType.AnnounceCancel
    case 0x11n:
      return ControlMessageType.SubscribeAnnounces
    case 0x12n:
      return ControlMessageType.SubscribeAnnouncesOk
    case 0x13n:
      return ControlMessageType.SubscribeAnnouncesError
    case 0x14n:
      return ControlMessageType.UnsubscribeAnnounces
    default:
      throw new Error(`Invalid ControlMessageType: ${v}`)
  }
}

/**
 * @public
 * Error codes for Announce control messages.
 */
export enum AnnounceErrorCode {
  InternalError = 0x0,
  Unauthorized = 0x1,
  Timeout = 0x2,
  NotSupported = 0x3,
  Uninterested = 0x4,
  MalformedAuthToken = 0x10,
  UnknownAuthTokenAlias = 0x11,
  ExpiredAuthToken = 0x12,
}

/**
 * Converts a bigint value to an AnnounceErrorCode enum.
 * @param v - The bigint value.
 * @returns The corresponding AnnounceErrorCode.
 * @throws Error if the value is not a valid announce error code.
 */
export function announceErrorCodeFromBigInt(v: bigint): AnnounceErrorCode {
  switch (v) {
    case 0x0n:
      return AnnounceErrorCode.InternalError
    case 0x1n:
      return AnnounceErrorCode.Unauthorized
    case 0x2n:
      return AnnounceErrorCode.Timeout
    case 0x3n:
      return AnnounceErrorCode.NotSupported
    case 0x4n:
      return AnnounceErrorCode.Uninterested
    case 0x10n:
      return AnnounceErrorCode.MalformedAuthToken
    case 0x11n:
      return AnnounceErrorCode.UnknownAuthTokenAlias
    case 0x12n:
      return AnnounceErrorCode.ExpiredAuthToken
    default:
      throw new Error(`Invalid AnnounceErrorCode: ${v}`)
  }
}

/**
 * @public
 * Filter types for subscription requests.
 */
export enum FilterType {
  NextGroupStart = 0x1,
  LatestObject = 0x2,
  AbsoluteStart = 0x3,
  AbsoluteRange = 0x4,
}

/**
 * Converts a bigint value to a FilterType enum.
 * @param v - The bigint value.
 * @returns The corresponding FilterType.
 * @throws Error if the value is not a valid filter type.
 */
export function filterTypeFromBigInt(v: bigint): FilterType {
  switch (v) {
    case 0x1n:
      return FilterType.NextGroupStart
    case 0x2n:
      return FilterType.LatestObject
    case 0x3n:
      return FilterType.AbsoluteStart
    case 0x4n:
      return FilterType.AbsoluteRange
    default:
      throw new Error(`Invalid FilterType: ${v}`)
  }
}

/**
 * @public
 * Fetch request types for MOQT protocol.
 */
export enum FetchType {
  StandAlone = 0x1,
  Relative = 0x2,
  Absolute = 0x3,
}

/**
 * Converts a bigint value to a FetchType enum.
 * @param v - The bigint value.
 * @returns The corresponding FetchType.
 * @throws CastingError if the value is not a valid fetch type.
 */
export function fetchTypeFromBigInt(v: bigint): FetchType {
  switch (v) {
    case 0x1n:
      return FetchType.StandAlone
    case 0x2n:
      return FetchType.Relative
    case 0x3n:
      return FetchType.Absolute
    default:
      throw new CastingError('fetchTypeFromBigInt', 'bigint', 'FetchType', `Invalid FetchType:${v}`)
  }
}

/**
 * @public
 * Group ordering options for object delivery.
 */
export enum GroupOrder {
  Original = 0x0,
  Ascending = 0x1,
  Descending = 0x2,
}

/**
 * Converts a number value to a GroupOrder enum.
 * @param v - The number value.
 * @returns The corresponding GroupOrder.
 * @throws CastingError if the value is not a valid group order.
 */
export function groupOrderFromNumber(v: number): GroupOrder {
  switch (v) {
    case 0x0:
      return GroupOrder.Original
    case 0x1:
      return GroupOrder.Ascending
    case 0x2:
      return GroupOrder.Descending
    default:
      throw new CastingError('groupOrderFromNumber', 'number', 'GroupOrder', `Invalid GroupOrder: ${v}`)
  }
}

/**
 * @public
 * Error codes for Subscribe control messages.
 */
export enum SubscribeErrorCode {
  InternalError = 0x0,
  Unauthorized = 0x1,
  Timeout = 0x2,
  NotSupported = 0x3,
  TrackDoesNotExist = 0x4,
  InvalidRange = 0x5,
  RetryTrackAlias = 0x6,
  MalformedAuthToken = 0x10,
  UnknownAuthTokenAlias = 0x11,
  ExpiredAuthToken = 0x12,
}

/**
 * Converts a bigint value to a SubscribeErrorCode enum.
 * @param v - The bigint value.
 * @returns The corresponding SubscribeErrorCode.
 * @throws Error if the value is not a valid subscribe error code.
 */
export function subscribeErrorCodeFromBigInt(v: bigint): SubscribeErrorCode {
  switch (v) {
    case 0x0n:
      return SubscribeErrorCode.InternalError
    case 0x1n:
      return SubscribeErrorCode.Unauthorized
    case 0x2n:
      return SubscribeErrorCode.Timeout
    case 0x3n:
      return SubscribeErrorCode.NotSupported
    case 0x4n:
      return SubscribeErrorCode.TrackDoesNotExist
    case 0x5n:
      return SubscribeErrorCode.InvalidRange
    case 0x6n:
      return SubscribeErrorCode.RetryTrackAlias
    case 0x10n:
      return SubscribeErrorCode.MalformedAuthToken
    case 0x11n:
      return SubscribeErrorCode.UnknownAuthTokenAlias
    case 0x12n:
      return SubscribeErrorCode.ExpiredAuthToken
    default:
      throw new Error(`Invalid SubscribeErrorCode: ${v}`)
  }
}

/**
 * @public
 * Error codes for Fetch control messages.
 */
export enum FetchErrorCode {
  InternalError = 0x0,
  Unauthorized = 0x1,
  Timeout = 0x2,
  NotSupported = 0x3,
  TrackDoesNotExist = 0x4,
  InvalidRange = 0x5,
  NoObjects = 0x6,
  InvalidJoiningRequestId = 0x7,
  UnknownStatusInRange = 0x8,
  MalformedTrack = 0x9,
  MalformedAuthToken = 0x10,
  ExpiredAuthToken = 0x12,
}

/**
 * Converts a bigint value to a FetchErrorCode enum.
 * @param v - The bigint value.
 * @returns The corresponding FetchErrorCode.
 * @throws CastingError if the value is not a valid fetch error code.
 */
export function fetchErrorCodeFromBigInt(v: bigint): FetchErrorCode {
  switch (v) {
    case 0x0n:
      return FetchErrorCode.InternalError
    case 0x1n:
      return FetchErrorCode.Unauthorized
    case 0x2n:
      return FetchErrorCode.Timeout
    case 0x3n:
      return FetchErrorCode.NotSupported
    case 0x4n:
      return FetchErrorCode.TrackDoesNotExist
    case 0x5n:
      return FetchErrorCode.InvalidRange
    case 0x6n:
      return FetchErrorCode.NoObjects
    case 0x7n:
      return FetchErrorCode.InvalidJoiningRequestId
    case 0x8n:
      return FetchErrorCode.UnknownStatusInRange
    case 0x9n:
      return FetchErrorCode.MalformedTrack
    case 0x10n:
      return FetchErrorCode.MalformedAuthToken
    case 0x12n:
      return FetchErrorCode.ExpiredAuthToken
    default:
      throw new CastingError('fetchErrorCodeFromBigInt', 'number', 'FetchErrorCode', 'Invalid fetch error code')
  }
}

/**
 * @public
 * Status codes for track status responses.
 */
export enum TrackStatusCode {
  InProgress = 0x00,
  DoesNotExist = 0x01,
  NotYetBegun = 0x02,
  Finished = 0x03,
  RelayUnavailable = 0x04,
}

/**
 * Converts a bigint value to a TrackStatusCode enum.
 * @param v - The bigint value.
 * @returns The corresponding TrackStatusCode.
 * @throws Error if the value is not a valid track status code.
 */
export function trackStatusCodeFromBigInt(v: bigint): TrackStatusCode {
  switch (v) {
    case 0x00n:
      return TrackStatusCode.InProgress
    case 0x01n:
      return TrackStatusCode.DoesNotExist
    case 0x02n:
      return TrackStatusCode.NotYetBegun
    case 0x03n:
      return TrackStatusCode.Finished
    case 0x04n:
      return TrackStatusCode.RelayUnavailable
    default:
      throw new Error(`Invalid TrackStatusCode: ${v}`)
  }
}

/**
 * @public
 * Error codes for SubscribeAnnounces control messages.
 */
export enum SubscribeAnnouncesErrorCode {
  InternalError = 0x0,
  Unauthorized = 0x1,
  Timeout = 0x2,
  NotSupported = 0x3,
  NamespacePrefixUnknown = 0x4,
  NamespacePrefixOverlap = 0x5,
  MalformedAuthToken = 0x10,
  UnknownAuthTokenAlias = 0x11,
  ExpiredAuthToken = 0x12,
}

/**
 * Converts a bigint value to a SubscribeAnnouncesErrorCode enum.
 * @param v - The bigint value.
 * @returns The corresponding SubscribeAnnouncesErrorCode.
 * @throws Error if the value is not a valid subscribe announces error code.
 */
export function subscribeAnnouncesErrorCodeFromBigInt(v: bigint): SubscribeAnnouncesErrorCode {
  switch (v) {
    case 0x0n:
      return SubscribeAnnouncesErrorCode.InternalError
    case 0x1n:
      return SubscribeAnnouncesErrorCode.Unauthorized
    case 0x2n:
      return SubscribeAnnouncesErrorCode.Timeout
    case 0x3n:
      return SubscribeAnnouncesErrorCode.NotSupported
    case 0x4n:
      return SubscribeAnnouncesErrorCode.NamespacePrefixUnknown
    case 0x5n:
      return SubscribeAnnouncesErrorCode.NamespacePrefixOverlap
    case 0x10n:
      return SubscribeAnnouncesErrorCode.MalformedAuthToken
    case 0x11n:
      return SubscribeAnnouncesErrorCode.UnknownAuthTokenAlias
    case 0x12n:
      return SubscribeAnnouncesErrorCode.ExpiredAuthToken
    default:
      throw new Error(`Invalid SubscribeAnnouncesErrorCode: ${v}`)
  }
}

/**
 * @public
 * Status codes for SubscribeDone control messages.
 */
export enum SubscribeDoneStatusCode {
  InternalError = 0x0,
  Unauthorized = 0x1,
  TrackEnded = 0x2,
  SubscriptionEnded = 0x3,
  GoingAway = 0x4,
  Expired = 0x5,
  TooFarBehind = 0x6,
}

/**
 * Converts a bigint value to a SubscribeDoneStatusCode enum.
 * @param v - The bigint value.
 * @returns The corresponding SubscribeDoneStatusCode.
 * @throws Error if the value is not a valid subscribe done status code.
 */
export function subscribeDoneStatusCodeFromBigInt(v: bigint): SubscribeDoneStatusCode {
  switch (v) {
    case 0x0n:
      return SubscribeDoneStatusCode.InternalError
    case 0x1n:
      return SubscribeDoneStatusCode.Unauthorized
    case 0x2n:
      return SubscribeDoneStatusCode.TrackEnded
    case 0x3n:
      return SubscribeDoneStatusCode.SubscriptionEnded
    case 0x4n:
      return SubscribeDoneStatusCode.GoingAway
    case 0x5n:
      return SubscribeDoneStatusCode.Expired
    case 0x6n:
      return SubscribeDoneStatusCode.TooFarBehind
    default:
      throw new Error(`Invalid SubscribeDoneStatusCode: ${v}`)
  }
}
